Pass new method in CollectServoSizes to get size of ObjectPrivateVisitor

aeweston98 opened the pull request:
<!-- Please describe your changes on the following line: -->
These changes are the servo changes to allow the measurement of DOM objects with the ObjectPrivateVisitor API. Two new functions have been added MemoryReporter and get_size, which are passed as function pointers to two updated mozjs functions. Currently these changes rely on a branch version of mozjs, which has an [open pull request](https://github.com/servo/rust-mozjs/pull/409).

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #7084 (github issue number if applicable).

<!-- Either: -->
- [ ] There are tests for these changes OR
- [x] These changes do not require tests because _____

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/20430)
<!-- Reviewable:end -->

aeweston98 committed:
Pass new method in CollectServoSizes for accurate DOM heap use reporting

highfive commented:
Thanks for the pull request, and welcome! The Servo team is excited to review your changes, and you should hear from @nox (or someone else) soon.

highfive commented:
Heads up! This PR modifies the following files:
 * @asajeffrey: components/script/lib.rs, components/script/Cargo.toml, components/script/script_thread.rs, components/script/script_runtime.rs
 * @fitzgen: components/script/lib.rs, components/script/Cargo.toml, components/script/script_thread.rs, components/script/script_runtime.rs
 * @kichjang: components/script/lib.rs, components/script/Cargo.toml, components/script/script_thread.rs, components/script/script_runtime.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!
* These commits modify script code, but no tests are modified. Please consider adding a test!

jdm commented on components/script/lib.rs:
```diff
@@ -199,6 +202,21 @@ pub fn init_service_workers(sw_senders: SWManagerSenders) {
 
+#[no_mangle]
+#[allow(unsafe_code)]
+pub unsafe extern "C" fn MemoryReporter(obj: *mut JSObject) -> bool {
```
Let's call this `is_dom_object`.

jdm commented on components/script/lib.rs:
```diff
@@ -199,6 +202,21 @@ pub fn init_service_workers(sw_senders: SWManagerSenders) {
+    if obj.is_null() {
+        return false;
+    }
+    if is_platform_object(obj) || is_dom_proxy(obj) {
```
This can be simplified to:
```rust
pub unsafe extern "C" fn MemoryReporter(obj: *mut JSObject) -> bool {
    !obj.is_null() && (is_platform_object(obj) || is_dom_proxy(obj))
}
```

jdm commented on components/script/lib.rs:
```diff
@@ -207,6 +225,8 @@ pub fn init() {
         // bindings to implement JS proxies.
         RegisterBindings::RegisterProxyHandlers();
+
+        js::glue::InitializeMemoryReporter(serde::export::Some(MemoryReporter));
```
There should be no need for `serde::export::` here.

jdm commented on components/script/script_runtime.rs:
```diff
@@ -312,14 +316,38 @@ pub unsafe fn new_rt_and_cx() -> Runtime {
+#[no_mangle]
+#[allow(unsafe_code)]
+pub unsafe extern "C" fn get_size(obj: *mut JSObject) -> usize {
+        let dom_class_opt = get_dom_class(obj);
```
This function looks like it's over-indented by 4 spaces.

jdm commented on components/script/script_runtime.rs:
```diff
@@ -312,14 +316,38 @@ pub unsafe fn new_rt_and_cx() -> Runtime {
+pub unsafe extern "C" fn get_size(obj: *mut JSObject) -> usize {
+        let dom_class_opt = get_dom_class(obj);
+
+        match dom_class_opt {
```
We should be able to do `match get_dom_class(obj) {`.

jdm commented on components/script/script_runtime.rs:
```diff
@@ -312,14 +316,38 @@ pub unsafe fn new_rt_and_cx() -> Runtime {
+
+        match dom_class_opt {
+            Ok(v) => {
+                let pass_ptr = private_from_object(obj) as *const c_void;
```
Let's call this `dom_object`.

jdm commented on components/script/script_runtime.rs:
```diff
@@ -312,14 +316,38 @@ pub unsafe fn new_rt_and_cx() -> Runtime {
+        match dom_class_opt {
+            Ok(v) => {
+                let pass_ptr = private_from_object(obj) as *const c_void;
+                if !(pass_ptr.is_null()) {
```
Let's write this as:
```rust
if dom_object.is_null() {
    return 0;
}
let mut ops = ...;
```
```

jdm commented on components/script/script_runtime.rs:
```diff
@@ -312,14 +316,38 @@ pub unsafe fn new_rt_and_cx() -> Runtime {
+                let pass_ptr = private_from_object(obj) as *const c_void;
+                if !(pass_ptr.is_null()) {
+                    let mut ops = MallocSizeOfOps::new(::servo_allocator::usable_size, None, None);
+                    let _result = (v.malloc_size_of)(&mut ops, pass_ptr);
```
We can return this directly instead of storing it in a local variable.

jdm commented on components/script/script_runtime.rs:
```diff
@@ -312,14 +316,38 @@ pub unsafe fn new_rt_and_cx() -> Runtime {
         let rt = JS_GetRuntime(cx);
         let mut stats = ::std::mem::zeroed();
-        if CollectServoSizes(rt, &mut stats) {
+        if CollectServoSizes(rt, &mut stats, serde::export::Some(get_size)) {
```
We should be able to remove the `serde::export::`.

jdm commented on components/script/script_thread.rs:
```diff
@@ -1581,32 +1579,7 @@ impl ScriptThread {
-        // malloc_enclosing_size_of function.
-        let mut ops = MallocSizeOfOps::new(::servo_allocator::usable_size, None, None);
-
-        for (_, document) in self.documents.borrow().iter() {
```
Let's keep the code that modifies path_seg.

jdm reviewed (changes requested):
Please squash these commits as well.

jdm commented:
I'm excited to see these changes almost ready!

aeweston98 commented:
@jdm I rebased from the latest changes in master, updated according to your comments and squashed the commits. I saw one flag in the change view of cargo.toml which says no newline at end of file, is this something I need to change?

jdm commented:
No, that should be fine.

jdm commented on components/script/script_runtime.rs:
```diff
@@ -312,14 +315,33 @@ pub unsafe fn new_rt_and_cx() -> Runtime {
 
+#[no_mangle]
+#[allow(unsafe_code)]
+pub unsafe extern "C" fn get_size(obj: *mut JSObject) -> usize {
```
nit: no need for `pub`.

jdm commented on components/script/script_thread.rs:
```diff
@@ -1581,30 +1579,11 @@ impl ScriptThread {
     fn collect_reports(&self, reports_chan: ReportsChan) {
         let mut path_seg = String::from("url(");
-        let mut dom_tree_size = 0;
         let mut reports = vec![];
```
This is unused now.

aeweston98 replied:
I am interpreting this to mean that since we don't add any dom-tree entries to the reports vector anymore we can use the return of get_reports directly. ie. `reports_chan.send(get_reports(self.get_cx(), path_seg));` Is that what you meant?

jdm replied:
Oh, my mistake. I was skimming the changes and didn't look closely enough. You can ignore this comment.

jdm commented on components/script/script_thread.rs:
```diff
@@ -1581,30 +1579,11 @@ impl ScriptThread {
-            dom_tree_size += malloc_size_of_including_self(&mut ops, document.window());
-
-            if reports.len() > 0 {
-                path_seg.push_str(", ");
```
We're missing the joining `, ` now. Let's do the following instead of this loop:
```
let documents = self.documents.borrow();
let urls = itertools::join(documents.values().map(|d| d.url().as_str()), ", ")
let path_seg = format!("url({})", urls);
```
(you'll need to add an `extern crate itertools` to script/lib.rs and add itertools to script/Cargo.toml, too)

aeweston98 replied:
This change results in a compilation error `no method named values found for type std::cell::Ref<'_, script_thread::Documents>`. I looked at the impl for Documents, but was not able to fix the problem. I will continue working on it tomorrow.

jdm replied:
Oh, I made the wrong assumption about the type of self.documents. You can use `documents.iter().map(|(_, d)| d.url().as_str())` instead.

aeweston98 replied:
When I put this in as is, it results in a compilation error on d.url() `borrowed value does not live long enough`. I think this is caused by as_str() returning a &str. I replaced as_str() with to_string() and it compiled and I tested it and the colon separated list of urls shows up. Please let me know if that change is ok.

jdm replied:
Looks good to me!

jdm commented on components/script/lib.rs:
```diff
@@ -200,6 +202,12 @@ pub fn init_service_workers(sw_senders: SWManagerSenders) {
 
+#[no_mangle]
+#[allow(unsafe_code)]
+pub unsafe extern "C" fn is_dom_object(obj: *mut JSObject) -> bool {
```
nit: no need for `pub`.

aeweston98 commented:
sorry about that extra commit, in process of rebasing from servo/servo master

jdm commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 60dfd66 has been approved by `jdm`

<!-- @bors-servo r=jdm 60dfd6635879608af10a141c9dca8ea03588402b -->

bors-servo commented:
:hourglass: Testing commit 60dfd6635879608af10a141c9dca8ea03588402b with merge 4efeda61fea8d1ec983dd58fde7918e794fcd354...

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/11468)

aeweston98 commented:
@jdm It was the lack of EOF in Cargo.toml which caused test-tidy to fail, fixed now. Also updated to latest changes in master.

jdm commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 7f7fc91 has been approved by `jdm`

<!-- @bors-servo r=jdm 7f7fc917586fa5ea9c0681f7a479c7ec8ad69019 -->

bors-servo commented:
:hourglass: Testing commit 7f7fc917586fa5ea9c0681f7a479c7ec8ad69019 with merge 339146ae1566fc35e47ccc4fba1471a9123d039d...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/11162), [arm32](http://build.servo.org/builders/arm32/builds/11397), [arm64](http://build.servo.org/builders/arm64/builds/11432), [linux-dev](http://build.servo.org/builders/linux-dev/builds/11469), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/8356), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/8366), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/11527), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/3920), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/3948), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/8195), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/8262), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/4020), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/3969), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/6861)
Approved by: jdm
Pushing 339146ae1566fc35e47ccc4fba1471a9123d039d to master...
